### ISSUE TITLE
Share_ID no longer generating

### ISSUE THREAD

**Issue by jpcsupplies** [MPOS/php-mpos#2784]
Had a wierd issue occur last weekend.  No changes have been made to the pool prior for weeks.  But as of Sunday 26th April the share_id is no longer populating in the blocks table, nor is shares, worker name account id etc, and accounted remains at 0.  Everything else seems to populate fine.

Share ID halted at about 1265732 which is not a particularly high number, and it does not appear to have rolled over.
Below is an example of the error log it generated prior to attempting to manually fix the ID.
`2020-04-30 10:12:17 - ERROR --> E0001: The block with height 1265732 found share 2308 which is < than 61834663 of block 1264115.
2020-04-30 10:12:17 - CRIT --> E0002: Failed to fetch both offending shares 2308 and 61834663. Block height: 1265732
2020-04-30 10:20:04 - CRIT --> Cronjob is currently disabled due to errors, use -f option to force running cron.`

This results in a record as pictured every time a miner hits a block:
![not increment](https://user-images.githubusercontent.com/13939039/80719087-50ea7f80-8b3e-11ea-93c0-8768c06a4673.jpg)

I've checked logs and users, and nothing odd stands out except one user the week prior somehow managed to mine over 338 blocks over about 12 days - leaving their pool balance at over 2000 with no autopayout threshold.

So far ive updated MPOS, updated the kernel, updated stratum, updated the coin daemon, manually inserted share ID's and flipped accounting.. which gets all the crons working again until the problem occurs again once they hit the next block.

Unless it is some sort of exploit related to the user mining 338 blocks which broke something - in which case updating MPOS in theory should have resolved it - it seems to be some sort of unusual bug.

**Comment by jpcsupplies** [comment]
I've gone through logs and the only anomaly I can find is in the twistd stratum-mining logs around the time of the issue show some bots spewing RDP garbage into port 3333 occaisionally.

I switched the pool backend stratum proxy over to NOMP and it still appears to be generating null share_id and account_id/worker_name/shares in nomp too - which seems to point to the issue being in MPOS still.

**Comment by jpcsupplies** [comment]
So, I spun up a new pool, new mpos folder, new sql database, about the only thing I kept was the twisted stratum-mining which i configured to talk to the new sql database. It appeared to work for a while - meanwhile i restarted the machine,  Got things going and now 
"findblock Unable to fetch blocks upstream share"   check the sql database, and the nul shareID's are back.
`2020-05-16 7:40:03 - INFO --> Starting block share detection, this may take a while
2020-05-16 7:40:03 - INFO --> | Block ID |     Height |          Amount |   Share ID |                    Shares | Finder          | Worker          |               Type |
2020-05-16 7:40:03 - CRIT --> E0005: Unable to fetch blocks upstream share, aborted:Unable to find valid upstream share for block: 1246973
`

**Comment by jpcsupplies** [comment]
Tried nomp again also, but its pretty useless, the mpos documentation support is so half assed it may as well be an undocumented feature. Miners cannot even connect. Appears it cnnot detect if a complete blockchain copy exists despite its own log reporting it at 100% - so the stratum server never initialises.

**Comment by TheSerapher** [comment]
NOMP is not part of MPOS and hence not documented - use their documentation to get miners to connect.

Your coin may just not be properly supported by MPOS or startum. There are a few mechanisms that will assist MPOS in finding a proper share for a block and if nothing matches it can even take a random share.

It might have been helpful for others to let them know what coin you are trying to mine, if others have the same issue they may be able to help.

**Comment by jpcsupplies** [comment]
Sure thing, the coin is a litecoin based scrypt coin called Marscoin.  
The MPOS pool has been minnig it no issues since 2015, it was only about 14 days ago, that it started throwing the issue.

**Comment by jpcsupplies** [comment]
btw by NOMP documentation i meant theirs not yours :)

**Comment by TheSerapher** [comment]
But if everything worked fine until now and nothing has changed in MPOS since then, why would the error causing this be somewhere in that old code? It’s also possible that the coin has changed something which is causing this. Probably somewhere in Stratum.

I’d start looking there especially since other pools seem also to be affected by whatever this is.

**Comment by jpcsupplies** [comment]
Thats the odd part - up until the fault, nothing has changed. The coin daemon was the same version, a newer client was released about a year back but it was the same protocol and had been working.     When the error occured miners still got data, and could still hit blocks, but when they hit blocks share ID/share was null.  If i manually set a share ID etc like in your documentation, it would all run on its merry way until the next block was hit, then splat null again.

**Comment by TheSerapher** [comment]
Hard to say what it is then - but what seems to be the case: The share that found a block was maybe inside the range of another block.

Did you notice anything special about these faulty blocks? Maybe each followed very quickly by pure luck?

**Comment by jpcsupplies** [comment]
The only thought that occured to me - was if twisted stratum-mining accessed an external resource somewhere, outside of the coindaemon - and that was no longer available.  
This occurs to me as I've also seen the occaisional E0007 out of order share autofix and E0005 unable to fetch upstream, and in rare cases an E0062 most of these usually fix themself as they were caused by internet connectivity issues or power outages and just required a -f on the crons to fix once connectivity was restored.

But this E0001 and E0002 seems to be a fatal error, even manually keying in data to get it moving again returns to it. So either I am chosing poorly my manual data, or I am missing something else here.

**Comment by TheSerapher** [comment]
But if it was running fine for years and suddenly starts acting up, even with a brand new setup and let it run a while, something must have changed. Sorry that I can't be of more help but often these errors are caused by stratum or coin issues and working around them with giving them a share ID is the only way MPOS can fix it - since there really is no matching Share ID available.

**Comment by jpcsupplies** [comment]
Block discovery by speed luck or brute force hasnt been an issue previously, the weeks prior somone hit about 2000 coin worth, in a week or so, which at the current reward of 6 would have been pushing some pretty heavy hash rates. Miners regularly ran up to 500mh on the pool each

**Comment by TheSerapher** [comment]
> Block discovery by speed luck or brute force hasnt been an issue previously, the weeks prior somone hit about 2000 coin worth, in a week or so, which at the current reward of 6 would have been pushing some pretty heavy hash rates. Miners regularly ran up to 500mh on the pool each

Yeah, MPOS has no issues with speedy block rates either, I have run tests with blocks popping up within seconds (and even faster 😉). But as long as Straum properly inserts all shares in proper order there is no issue with this. I think there was even a patch for it a long time ago that ensured shares are added in proper order.

**Comment by jpcsupplies** [comment]
can you give any tips on manually setting shareids in case I am just screwing it up (although that doesnt explain how rerolling a pool gave a similar issue :/)  the last time i tried was using phpmyadmin (see screenshot above for example bad record)

Or give me a better idea of the order of information flow -
what info is being set by MPOS and what is coming from stratum-mining or the coin demon.

The sql seems to show height, hash, confirmations, amount, timestamp, all registered fine, but then  0, null or unknown on accounted (0), account_id (null), workername(unknown), shares(null), and share_id(null) 

The _next_ issue that may be a problem, which may complicate any fixes at this point is out of desperation trying to fix the original issue I tried updating ubuntu from 14 to 16, and upgraded MPOS to the latest. This required a recompile of the coindemon. stratum-mining had not changed since first install and still worked as is (tho reportedly its installer is broken under 16 thanks to depreciated stuff in python!?) the pool reroll was under 16.. so ye.

**Comment by TheSerapher** [comment]
MPOS only handles the share information that Stratum inserted into the table. It takes care of Payouts while Stratum takes care of the work units and shares.

Manually choosing a share ID via PHPMyAdmin is fine, just make sure you are not using yet another share that solved a block or beyond the share of a newly found block. Once assigned MPOS will calculate the payouts.

The upgrade sure makes it harder but depending how far behind you were regarding MPOS it should not cause these issues. That part hasn't changed in a long time.

**Comment by jpcsupplies** [comment]
Ok so that is helpful then, in theory the null data is coming out of twisted stratum-mining then?
Are you aware of anything else than nomp and twisted stratum-mining that could take on that role?

For my edited values here is a few i tried with little sucess -
For share ID, I started with the last working one, and then incremented by 1 each height, that got the pool moving again but the nextblock it choked again. then For the next few nulls i just used the same id, and that didnt help either, also tried entering in a worker and account at one point - which showed some results on the pool info page but didnt help either.

Ill attach an example screen of my inserted values starting with the last 3 valid and a page of nulls, and my attempts at edited values.

I wonder if i simply deleted all the invalid values back to the last valid one if that would work too? probably break cross references to other tables tho :/

So everything from valid line up generated no issue.
Everything below is data that came up null and I edited or otherwise tried to get MPOS to skip past.
![sampledata](https://user-images.githubusercontent.com/13939039/82189796-5532dd00-9933-11ea-8064-e7cb2a789758.jpg)

**Comment by HashUnlimited** [comment]
There are 2 possibilities:
a.) You have enabled anonymous accounts in your stratum. NOMP got a setting for that for example but MPOS can't handle it.

b.) You are mining to the same address used in the pool from another place. That's not allowed, MPOS is assuming that ONLY the wallet connected to the pool owns the privkey (which is the way it should be). Imo this is a flaw in MPOS as anyone can break the pool by mining to this address but at least it comes at the price of the blocks.

**Comment by TheSerapher** [comment]
If you deleted blocks it will re-discover them as new and try to do the share calculations, so better not touch that :D

And if those are the blocks found, dont skip any with NULL, all blocks require valid shares. If you don't have any more shares to use as fake valid without risking picking on that may already be inside a new block, just set accounted to 1 and MPOS will assume they are paid out. You can then use the pool bonus payout feature to pay these out to miners - but be careful to also disable it once the bonus equals to the left over coins of the blocks not properly paid out. This is easier and just as effective if you prefer to not deal with NULL share IDs.

And @HashUnlimited gave some good hints too, make sure you are mining properly :)

**Comment by jpcsupplies** [comment]
Hmm, at the end there one worker had set his MPOS account to show in the pool info page as anonymous, but that was an MPOS setting not in the stratum. When logged in as admin I could see their real name fine.  At that point I was still running twisted stratum-mining anyway. 

So @TheSerapher you are saying if i leave it all null but flip accounting instead that should have it roll past too?   I probably wouldnt need to use bonus - in the past I've manually paid out miners from my own funds - since there was at most the same 2 or 3 guys mining anyway and I could work out roughly how much they earnt with a little extra as an apology anyway.

@HashUnlimited 
But that brings up other interesting possibilities -
1: Possibility: 
I screwed up: I am still using the same wallet address in both attempts at re-rolling the pool - pool 2 might have choked because the defined address already had transactions.  In which case the reroll could have worked if I used a free address instead?

2: Possibility: 
Somone doing this deliberately:  Mining into the same address - there was as stated some.. odd things happening in the logs leading up to the failure.   stratum-mining is likely common between all the pools that choked - and likely use a similar configuration.   Could somone have used some sort of malformed miner query  to connect via stratum port using odd options or used one of those anon pools where your worker name is the address you want it to pay out to - then deliberately mined into it to break the pool - sounds like a potential DoS weakness - and given all 3 pools dropped at a similar time - seems at least possible.
The rate at which null blocks appeared was at the time faster than the discovery rate of the active miners at the time - and a new mystery miner a week prior had pushed such crazy high hask into the pool they generated 2000 coiins, which they never withdrew and never logged in again.. and the pool dropped a week later. Trying to discover the pool wallet address perhaps?

**Comment by HashUnlimited** [comment]
Possibility 1 looks very familiar. When doing a re-setup of any pool, always use a new address. Usually it's safe to use a backup of the old DB unless it's manually been messed with - just make sure the new key contains enough balance to pay the open bills.

The rest sounds like a lot of speculation to me. I assume you are talking about Ahmed's python which isn't maintained afaik since BTC 0.14 or so - if it works with your coin, chances are that the codebase itself got unfixed issues like double spend wormholes (CVE-2018-17144) and someone is going after it.

**Comment by jpcsupplies** [comment]
@HashUnlimited 
On point 1 - ill try a reroll of the reroll with a shiny new wallet address on the same wallet so there is already funds as a whole there.  (RPC send just sends from any available address with funds, which is I beleive how MPOS initiates payouts?)

On point 2:
the twistd stratum-mining I use is indeed Ahmed's -
IE 
git clone github.com/Tydus/litecoin_scrypt.git
git clone github.com/ahmedbodi/stratum-mining.git
git clone github.com/ahmedbodi/stratum.git
etc etc

Can you suggest an alternative, the nomp I tried seemed pretty broken, lacked useful documentation (at least where MPOS was concerned)  and looked just as abandoned as ahmed's.   ?

@TheSerapher  So if someone gets a hold of your pool  address and deliberately mines into it - that potentially breaks a pool?   I suppose I could change my pool address - but depending how someone locates the address this might not help much if they have another go later.. and I assume that would cause even more carnage in the sql database - any way MPOS could be tweaked to make that weakness more difficult to exploit ?

**Comment by TheSerapher** [comment]
MPOS uses the address you have configured in it for payouts, none other. You may have to transfer funds to the new address.

I am not aware that one can just mine into a pool address willy nilly and cause another pool to break. No clue though since that's something of a coin implementation thing. MPOS has no means of knowing who mined into the address so all blocks are valid, even if there are no shares to it.

**Comment by HashUnlimited** [comment]
The payment processor is only using `sendtoaddress` or if enabled `sendmany` so the wallet itself doesn't care and choses the best UTXOs independently to send the payouts (be careful about small amounts and tx size limits).

I agree, NOMP is poorly maintained as well - no issue for coins with ancient codebase. If you got a newer one, you can use the development branch of NOMP, at least there we got push rights there (not on master) and updating it on RPC changes and so on where needed.

My personal approach (getting hold of your pool address isn't exactly rocket science and mining a few blocks is cheap) in general, especially on crapcoins, is to mine every block to a new address. Didn't publish this option in NOMP so far because that makes the MPOS wallet site look ugly and convincing the community to entirely remove the address info isn't done in a day.

**Comment by HashUnlimited** [comment]
Update (had to check on the process first, has been a while back): Do NOT use the same wallet, rather a clean one. findblock is looking for ALL CB transactions using `listsinceblock` which includes all tx we own the privkey for.

**Comment by jpcsupplies** [comment]
Ahh ok thanks @HashUnlimited  so I'll need to spawn a fresh wallet entirely, not just a fresh address? Which means ugh ill have to pay out the entire old wallet balance, create a clean one, deposit what I cleaned out back to the new address etc.. 
Actually not so bad I can just move the wallet profile folder around between stop/starts of the wallet daemon. Cut paste the new address, move it back pay the new address manually, stop rename.. restart etc lol.     Messy. But possible.  Probably even possible to import/export the miner accounts in the sql across.. i think. Have to double check the tables to make sure there isnt any secondary relationship tables.  Never could work out where it stores each miners balance in the sql.. tho. The utils scripts gave me an account dump and wallet balance so it has to be there somewhere.

(also: So the only stratum alternative to run MPOS is NOMP's ?
marscoin isnt particularly ancient it uses the same codebase as litecoin. So maybe I need the dev.  Its docs related to MPOS were contradictory and incomplete that I found. :(  no basic stuff like setting your ports, or actually enabling the mpos profile.   Once it was running it sees blockchain as 100% downloaded but still decided it was incomplete, so never starts the miner port. Which is what threw me.)

**Comment by HashUnlimited** [comment]
If your DB is OK, you don't need to pay out manually. MPOS is doing the job for you.

@TheSerapher probably this explains the thing above. The `listsinceblock` call returns all transactions which can be decoded by our local wallet (it's always been like that). The cron is looking for the coinbase:

    if ( $aData['category'] == 'generate' || $aData['category'] == 'immature' ) {

Which is has been needed by by the RPC accordingly before:

            if (wtx.IsCoinBase())
            {
                if (wtx.GetDepthInMainChain() < 1)
                    entry.pushKV("category", "orphan");
                else if (wtx.IsImmatureCoinBase())
                    entry.pushKV("category", "immature");
                else
                    entry.pushKV("category", "generate");
            }

As a conclusion, if any coinbase transaction pays to any privkey known by tho pool wallet, it will cause the cron to fail for a block without share.

**Comment by jpcsupplies** [comment]
So an update on this - there was another issue with a stalled blockchain, which is resolved now so I rerolled the pool as discussed, but using nomp instead of twistd.  I imported the old user list and worker list so existing miners could jump right in once the pool was running.

This resulted in the wierd situation that it scanned through every block from about 371246 and added them all, then randomly assigned workers as whoever was trying to mine at that moment- rolling them forward each run of cron and halting sometimes until next run eg
`
2020-10-23 2:41:56 - INFO --> |    17118 |     379037 |              50 |        269 |                       512 | [1] me    | me.8      |          any_share |

2020-10-23 2:41:57 - CRIT --> E0005: Unable to fetch blocks upstream share, aborted:Unable to find valid upstream share for block: 379095

2020-10-23 2:50:04 - INFO --> Starting block share detection, this may take a while

2020-10-23 2:50:04 - INFO --> | Block ID |     Height |          Amount |   Share ID |                    Shares | Finder          | Worker          |               Type |

2020-10-23 2:50:04 - INFO --> |    72608 |     379095 |          50.007 |        270 |                       512 | [1] me    | me.8      |          any_share |

2020-10-23 2:50:04 - INFO --> |    57742 |     379104 |              50 |        271 |                       512 | [1] me    | me.8      |          any_share |
`


Eventually it got to the end -
`
2020-10-24 1:55:56 - INFO --> |    92467 |    1266890 |            6.25 |      93233 |                        32 | [1] me    | me.8      |          any_share |

2020-10-24 1:55:57 - INFO --> |     7250 |    1266893 |            6.25 |      93234 |                       512 | [7] user1    | user1.1      |          any_share |

2020-10-24 1:55:58 - INFO --> |    13886 |    1266906 |            6.25 |      93235 |                       512 | [76] user2     | user2.user1  | 
 `


Once it finally worked its way to the current block it started throwing upstream share already assigned to previous block -
`
2020-10-25 0:17:53 - INFO --> | 6e3768c2fbc7038db... |         1273367 |       6.25 |            86 |                8.10295222 | 2020-10-24 21:09:4 |

2020-10-25 0:17:53 - INFO --> | db46b55e4d9adc9e2... |         1272985 |       6.25 |           468 |                8.52609978 | 2020-10-24 07:33:2 |

2020-10-25 0:17:53 - INFO --> | 1b24dda96b3d6ed3d... |         1273065 |       6.25 |           388 |               11.93803337 | 2020-10-24 10:12:5 |

2020-10-25 0:17:54 - INFO --> Starting block share detection, this may take a while

2020-10-25 0:17:54 - INFO --> | Block ID |     Height |          Amount |   Share ID |                    Shares | Finder          | Worker          |               Type |

2020-10-25 0:17:54 - CRIT --> Upstream share already assigned to previous block

2020-10-25 0:20:03 - CRIT --> Cronjob is currently disabled due to errors, use -f option to force running cron.

2020-10-25 0:23:10 - INFO --> Starting RPC block detecion, blocks are stored in Database

2020-10-25 0:23:10 - INFO --> | Blockhash            |          Height |     Amount | Confirmations |                Difficulty |               Time |

2020-10-25 0:23:10 - INFO --> | b0584ccae65d42384... |         1273453 |       6.25 |             1 |                10.0309713 | 2020-10-25 00:17:4 |

2020-10-25 0:23:10 - INFO --> Starting block share detection, this may take a while

2020-10-25 0:23:10 - INFO --> | Block ID |     Height |          Amount |   Share ID |                    Shares | Finder          | Worker          |               Type |

2020-10-25 0:23:10 - CRIT --> Upstream share already assigned to previous block
`


At this point I've no idea what to do next

**Comment by jpcsupplies** [comment]
Ok.   Follow up on this - for those with this issue:
I have done a complete pool re-build.   Wiped all the sql tables back to factory, after manually paying out to every miner any outstanding balance.  (move/delete database then run 
`sudo mysql -p -e "create database mpos"
sudo mysql -p mpos < sql/000_base_structure.sql`
to rebuild a fresh MPOS database.)

I created an entirely isolated coindaemon on another dedicated user account, updated everything to the latest ubuntu 16.X updates.  This leaves the original and its coinbalance intact for paying out miners on the old database without interfering with the re-rolled pool.

Pool sort of works now- still got error.
`- CRIT --> No shares found for this block, aborted: error`

This is a new error, so this is a forward step.  (it relates to SQL version see below)

Next-
Since any newer build of stratum-mining fails, fixing the already installed one was my only option. I have hacked `bitcoin_rpc.py`   to manually implement this change: https://github.com/Crypto-Expert/stratum-mining/commit/364e47d376f660bdf571a1931b9833448363be7b
(As a workaround for QT removing RPC support for solo mining and `getwork` in an upstream of the coin wallet code base.)

Now to fix the last `no shares found for block` error i noticed the sql tables were actually all correct this time (unlike the OP where data was missing) - it appears in my OCD in updating everything it also updated the SQL version.  So the fix here was to use SQL query As discussed here: https://github.com/MPOS/php-mpos/wiki/FAQ#is-mysql-57-supported  -
`mysql> set global sql_mode='STRICT_TRANS_TABLES,NO_ZERO_IN_DATE,NO_ZERO_DATE,ERROR_FOR_DIVISION_BY_ZERO,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION';`

So far everything is working as expected and i have a test miner to spin up useful data. Working ok.
